src/main/scala/is/hail/linalg/BlockMatrix.scala

```diff
@@ -426,14 +652,22 @@ class BlockMatrix(val blocks: RDD[((Int, Int), BDM[Double])],
         }
       }
     }
-    new BlockMatrix(blocks, blockSize, nRows, nCols)
+    new BlockMatrix(newBlocks, blockSize, nRows, nCols)
   }
 
-  def map4(dm2: M, dm3: M, dm4: M, op: (Double, Double, Double, Double) => Double): M = {
-    requireZippable(dm2)
-    requireZippable(dm3)
-    requireZippable(dm4)
-    val blocks = this.blocks.zipPartitions(dm2.blocks, dm3.blocks, dm4.blocks, preservesPartitioning = true) { (it1, it2, it3, it4) =>
+  def map4(bm2: M, bm3: M, bm4: M,
+    op: (Double, Double, Double, Double) => Double,
+    name: String = "operation",
+    reqDense: Boolean = true): M = {
+    if (reqDense) {
+      bm2.requireDense(name)
```

Comment:
do you also need `requireDense(name)` here?

Reply:
yes, good catch